src/http/request.cpp

```diff
@@ -82,59 +101,59 @@ void request_connect(Settings settings, Callback<Error, Var<Transport>> txp,
 }
 
 void request_send(Var<Transport> txp, Settings settings, Headers headers,
-                  std::string body, Callback<Error> callback) {
-    Request request;
-    Error error = request.init(settings, headers, body);
-    if (error) {
-        callback(error);
+                  std::string body, Callback<Error, Var<Request>> callback) {
+    request_maybe_send(Request::make(settings, headers, body), txp, callback);
+}
+
+void request_maybe_send(ErrorOr<Var<Request>> request, Var<Transport> txp,
+                        Callback<Error, Var<Request>> callback) {
+    if (!request) {
+        callback(request.as_error(), nullptr);
         return;
     }
-    // TODO: here we can simplify by using net::write()
-    txp->on_error([txp, callback](Error error) {
-        txp->on_error(nullptr);
-        txp->on_flush(nullptr);
-        callback(error);
-    });
-    txp->on_flush([txp, callback]() {
-        txp->on_error(nullptr);
-        txp->on_flush(nullptr);
-        callback(NoError());
-    });
     Buffer buff;
-    request.serialize(buff);
-    txp->write(buff);
+    (*request)->serialize(buff);
+    net::write(txp, buff, [=](Error err) {
+        callback(err, *request);
+    });
 }
 
 void request_recv_response(Var<Transport> txp,
                            Callback<Error, Var<Response>> cb,
                            Var<Reactor> reactor, Var<Logger> logger) {
     Var<ResponseParserNg> parser(new ResponseParserNg);
     Var<Response> response(new Response);
-    Var<bool> prevent_emit(new bool(false));
 
+    // Note: any parser error at this point is an exception catched by the
+    // connection code and routed to the error handler function below
     txp->on_data([=](Buffer data) { parser->feed(data); });
+
     parser->on_response([=](Response r) { *response = r; });
 
     // TODO: here we should honour the `ignore_body` setting
     parser->on_body([=](std::string s) { response->body += s; });
 
-    // TODO: we should improve the parser such that the transport forwards the
-    // "error" event to it and then the parser does the right thing, so that the
-    // code becomes less "twisted" here.
-
```

Comment:
`TODO` addressed by duplicating few lines of code and removing the `prevent_emit` hack